secio/README.md

```diff
@@ -0,0 +1,160 @@
+# SECIO 1.0.0
+
+> A stream security transport for libp2p. Streams wrapped by SECIO use secure
+> sessions to encrypt all traffic.
+
+## Authors
+
+- [Juan Benet](https://github.com/jbenet) (October, 2015)
+
+## Editors
+
+- [Cole Brown](https://github.com/bigs)
+- [Lars Gierth](https://github.com/lgierth)
+
+## Implementations
+
+- [js-libp2p-secio](https://github.com/libp2p/js-libp2p-secio)
+- [go-secio](https://github.com/libp2p/go-libp2p-secio)
+
+## Table of Contents
+
+- [Algorithm Support](#algorithm-support)
+
+## Algorithm Support
+
+SECIO allows participating peers to support a subset of the following
+algorithms.
+
+### Exhchanges
+
+- P-256
+- P-384
+- P-521
+
+### Ciphers
+
+- AES-256
+- AES-128
+- Blowfish
+
+### Hashes
+
+- SHA-256
+- SHA-512
+
+## Data Structures
+
+The SECIO wire protocol features two message types defined in the
+[protobuf description language](https://github.com/libp2p/go-libp2p-secio/blob/master/pb/spipe.proto).
+These two messages, `Propose` and `Exchange` are the only serialized types
+required to implement SECIO.
+
+## Protocol
+
+### Proposal Generation
+
+SECIO channel negotiation begins with a proposal phase. Both sides generate a
+proposal as follows:
+
+*Note: the public key should be serialized per the `Bytes` method from
+[go-libp2p-crypto](https://godoc.org/github.com/libp2p/go-libp2p-crypto#Key).*
+
+```
+Propose{
+  Rand: <16 secure random bytes>,
+  Pubkey: <public key bytes>,
+  Exchanges: <comma separated string of supported key exchanges>,
+  Ciphers: <comma separated string of supported ciphers>,
+  Hashes: <comma separated string of supported hashes>,
+}
+```
+
+Both peers serialize this message and send it over the wire. Upon deserializing
+their peer's message, they verify that the pubkey matches that described by the
+peer's peer ID (NOTE: this is sometimes only possible for the peer *initiating*
+the connection.) If the key doesn't match the peer ID, the peer can close the
+connection. If the peer doesn't know the remote peer's ID, they can compute and
+store the remote peer ID for later use.
+
+### Determining Roles and Algorithms
+
+Next, the peers use a deterministic formula to compute their roles in the coming
+exchanges. Each peer computes:
+
+```
+oh1 := sha256(concat(remotePeerPubKeyBytes, myNonce))
```

Comment:
This is the first use of the word "nonce". Does it refer to the `rand` field of the `Propose` message?

Reply:
Yes.